### PR TITLE
8311163: Parallel: Improve large object handling during evacuation

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change to ParallelGC young generation collection's handling
of large objArrays, to now use the infrastructure provided by JDK-8253237 and
JDK-8337709.  (That's the same infrastructure used by G1 young/mixed
collections.)

Testing: mach5 tier1-5

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311163](https://bugs.openjdk.org/browse/JDK-8311163): Parallel: Improve large object handling during evacuation (**Enhancement** - P4)

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) Review applies to [e1db9bca](https://git.openjdk.org/jdk/pull/20720/files/e1db9bcad41d82ce303ebfba051673d580e19539)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20720/head:pull/20720` \
`$ git checkout pull/20720`

Update a local copy of the PR: \
`$ git checkout pull/20720` \
`$ git pull https://git.openjdk.org/jdk.git pull/20720/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20720`

View PR using the GUI difftool: \
`$ git pr show -t 20720`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20720.diff">https://git.openjdk.org/jdk/pull/20720.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20720#issuecomment-2311334165)

### CODE REVIEW

**Comment by albertnetymk** [comment]
Using `-Xmx3g -XX:+UseParallelGC -XX:ParallelGCThreads=6 DelayInducer.java`, I can see large regression on my box (nproc=12). Have you seen sth similar?

```
# baseline
[0.002s][info][gc] Using Parallel
[1.252s][info][gc] GC(0) Pause Young (Allocation Failure) 381M->296M(479M) 50.696ms
[1.490s][info][gc] GC(1) Pause Young (Allocation Failure) 422M->422M(674M) 165.177ms
[1.864s][info][gc] GC(2) Pause Young (Allocation Failure) 673M->674M(925M) 212.106ms
[2.243s][info][gc] GC(3) Pause Young (Allocation Failure) 925M->925M(1428M) 229.982ms

# new
[0.005s][info][gc] Using Parallel
[5.504s][info][gc] GC(0) Pause Young (Allocation Failure) 381M->299M(479M) 203.409ms
[6.233s][info][gc] GC(1) Pause Young (Allocation Failure) 424M->425M(676M) 620.718ms
[7.594s][info][gc] GC(2) Pause Young (Allocation Failure) 676M->676M(928M) 1155.301ms
[8.920s][info][gc] GC(3) Pause Young (Allocation Failure) 927M->928M(1430M) 1119.259ms
```

**Comment by kimbarrett** [comment]
> Using `-Xmx3g -XX:+UseParallelGC -XX:ParallelGCThreads=6 DelayInducer.java`, I can see some regression (~10%) in avg gc-pause time on my box (nproc=12). Have you seen sth similar?

No, I'm not seeing anything like that. Indeed, I'm seeing strongly the
opposite. Note that my machine has nproc=32, but the results were similar with
-XX:ActiveProcessorCount=12.

```
old
[1.966s][info][gc] GC(0) Pause Young (Allocation Failure) 504M->387M(1930M) 744.238ms
[2.446s][info][gc] GC(1) Pause Young (Allocation Failure) 891M->882M(2283M) 236.505ms
[7.733s][info][gc] GC(2) Pause Young (Allocation Failure) 1686M->1669M(2668M) 1678.807ms
[9.423s][info][gc] GC(3) Pause Full (Allocation Failure) 2523M->89M(2421M) 45.843ms

new
[1.694s][info][gc] GC(0) Pause Young (Allocation Failure) 504M->397M(1930M) 469.406ms
[2.140s][info][gc] GC(1) Pause Young (Allocation Failure) 901M->892M(2283M) 224.942ms
[6.533s][info][gc] GC(2) Pause Young (Allocation Failure) 1682M->1669M(2668M) 761.521ms
[8.253s][info][gc] GC(3) Pause Full (Allocation Failure) 2509M->89M(2408M) 47.130ms
```

These results are pretty stable across 10-12 runs. Regardless of which variant
is used, the number and types of GCs is stable, as are the collection amounts.
And for a given variant the times are fairly stable.

For GC(0), the new variant is consistently faster, though the fastest old time
and the slowest new time are close or overlap.

For GC(1), the two variants are pretty similar.

For GC(2), the new variant is consistently more than a factor of 2 faster, and
the average difference is closer to 2.5x.

Digging a bit deeper, here are sample times for the Scavenge phase.
```
old
[7.849s][debug][gc,phases      ] GC(2) Scavenge 1928.903ms
new
[6.511s][debug][gc,phases      ] GC(2) Scavenge 774.372ms
```

I haven't found a way to get more detailed timing information there.  In
particular, I thought there was some logging in the TaskTerminator, but don't
see any now.

I rebuilt the variants with -DTASKQUEUE_STATS and ran with gc+task+stats=trace
logging, and here's a sample output:

```
[6.610s][trace][gc,task,stats] GC(2) thr       push      steal    chunked     chunks
[6.610s][trace][gc,task,stats] GC(2) --- ---------- ---------- ---------- ----------
[6.610s][trace][gc,task,stats] GC(2)   0     306340          0         19     306335
[6.610s][trace][gc,task,stats] GC(2)   1     300040          2          7     300041
[6.610s][trace][gc,task,stats] GC(2)   2     303588          1          0     303589
[6.610s][trace][gc,task,stats] GC(2)   3     334770          3          0     334771
[6.610s][trace][gc,task,stats] GC(2)   4     332980          3          0     332981
[6.610s][trace][gc,task,stats] GC(2)   5     436049          1          0     436050

new
[6.657s][trace][gc,task,stats] GC(2)     ----partial array----     arrays      array
[6.657s][trace][gc,task,stats] GC(2) thr       push      steal    chunked     chunks
[6.658s][trace][gc,task,stats] GC(2) --- ---------- ---------- ---------- ----------
[6.658s][trace][gc,task,stats] GC(2)   0     315502          2         18     315503
[6.658s][trace][gc,task,stats] GC(2)   1     314152          1          0     314153
[6.658s][trace][gc,task,stats] GC(2)   2     321092          2          7     321093
[6.658s][trace][gc,task,stats] GC(2)   3     406893          4          0     406893
[6.658s][trace][gc,task,stats] GC(2)   4     326710          1          0     326707
[6.658s][trace][gc,task,stats] GC(2)   5     329418          1          1     329418
```

The number of partial array steals is about the same.

So I don't yet know why the change is so beneficial for GC(2).

**Comment by tschatzl** [comment]
I did some runs with 1, 2, 4, 6, 12 and 25 parallel workers and I can't see regressions either; the changes seem to be better always. The only difference from your runs is that with `-Xmx3g` there are in total four young collections every time. The last one is the one showing the 2x+ speedup.

**Comment by tschatzl** [comment]
The old code seems to only ever push a single continuation task, while the new code pushes multiple (after the first one), and tries to keep them at `ParallelGCThreads`.

**Comment by albertnetymk** [comment]
Using another box (AMD), the improvement becomes clear.

```
## baesline

[0.003s][info][gc] Using Parallel
[1.464s][info][gc] GC(0) Pause Young (Allocation Failure) 512M->344M(1963M) 586.065ms
[2.163s][info][gc] GC(1) Pause Young (Allocation Failure) 857M->853M(2304M) 298.110ms
[8.208s][info][gc] GC(2) Pause Young (Allocation Failure) 1707M->1669M(2689M) 2986.103ms
[9.941s][info][gc] GC(3) Pause Full (Allocation Failure) 2516M->91M(2485M) 38.478ms

## new
[0.002s][info][gc] Using Parallel
[1.325s][info][gc] GC(0) Pause Young (Allocation Failure) 512M->355M(1963M) 415.916ms
[1.791s][info][gc] GC(1) Pause Young (Allocation Failure) 867M->858M(2304M) 212.690ms
[5.663s][info][gc] GC(2) Pause Young (Allocation Failure) 1700M->1669M(2689M) 821.355ms
[7.088s][info][gc] GC(3) Pause Full (Allocation Failure) 2510M->91M(2475M) 32.170ms
```

> So I don't yet know why the change is so beneficial for GC(2).

With `-Xlog:gc*=debug`, I can see an expansion during GC(2) -- I guess because the expansion operation is synchronous, not having any chunks in the task-queue essentially blocks other workers.

File: src/hotspot/share/gc/parallel/psPromotionManager.cpp [hunk 1]
```diff
@@ -277,37 +295,57 @@ template <class T> void PSPromotionManager::process_array_chunk_work(
...
-    start = 0;
-    int const actual_length = arrayOop(obj)->length();
-    arrayOop(old)->set_length(actual_length);
+    process_array_chunk_work<oop>(state->destination(), start, end);
   }
+  // Release reference to state, now that we're done with it.
+  _partial_array_state_allocator->release(_partial_array_state_allocator_index, state);
+}
 
+void PSPromotionManager::push_objArray(oop old_obj, oop new_obj, size_t obj_size) {
```
Comment (albertnetymk) [inline src/hotspot/share/gc/parallel/psPromotionManager.cpp:323]: `obj_size` seems unused.

**Review by albertnetymk — APPROVED** [review]

**Review by tschatzl — APPROVED** [review]
lgtm

File: src/hotspot/share/gc/parallel/psPromotionManager.cpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/share/gc/parallel/psPromotionManager.cpp]: I'd originally intended to compute the array_length from this argument, but that's kind of messy so I went
with the simpler approach of just fetching it from one of the objects.  I've removed the argument.

**Review by tschatzl — APPROVED** [review]
Still good.

**Comment by kimbarrett** [comment]
Thanks for reviews @albertnetymk and @tschatzl

**Pull request closed without merging**